pkg/controller/framework/controller_test.go

```diff
@@ -113,3 +113,60 @@ func Example() {
 	// b-controller
 	// c-framework
 }
+
+func ExampleInformer() {
+	// source simulates an apiserver object endpoint.
+	source := framework.NewFakeControllerSource()
+
+	// Let's do threadsafe output to get predictable test results.
+	outputSetLock := sync.Mutex{}
+	outputSet := util.StringSet{}
+
+	// Make a controller that immediately deletes anything added to it, and
+	// logs anything deleted.
+	_, controller := framework.NewInformer(
+		source,
+		&api.Pod{},
+		time.Millisecond*100,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				source.Delete(obj.(runtime.Object))
+			},
+			DeleteFunc: func(obj interface{}) {
+				key, err := framework.DeletionHandlingMetaNamespaceKeyFunc(obj)
+				if err != nil {
+					key = "oops something went wrong with the key"
+				}
+
+				// Record some output when items are deleted.
+				outputSetLock.Lock()
+				defer outputSetLock.Unlock()
+				outputSet.Insert(key)
+			},
+		},
+	)
+
+	// Run the controller and run it until we close stop.
+	stop := make(chan struct{})
+	controller.Run(stop)
```

Comment:
I figured.  My point was that, as I understand, the general Go pattern is
for the caller to create the goroutine, rather than a library function.
It's all perspective, i guess, but as a consumer of this library I would
expect to say `go controller.Run(stop)`  Not a huge deal.

On Tue, Apr 7, 2015 at 5:16 PM, Daniel Smith <notifications@github.com>
wrote:

> In pkg/controller/framework/controller_test.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6546#discussion_r27933403>
> :
>
> > +				key, err := framework.DeletionHandlingMetaNamespaceKeyFunc(obj)
> > +				if err != nil {
> > +					key = "oops something went wrong with the key"
> > +				}
> > +
> > +				// Record some output when items are deleted.
> > +				outputSetLock.Lock()
> > +				defer outputSetLock.Unlock()
> > +				outputSet.Insert(key)
> > +			},
> > +		},
> > +	)
> > +
> > +	// Run the controller and run it until we close stop.
> > +	stop := make(chan struct{})
> > +	controller.Run(stop)
>
> Run itself starts two goroutines.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6546/files#r27933403>
> .
>